lectures/mle.md

```diff
@@ -0,0 +1,249 @@
+numerical methods to solve for parameter estimates.
+
+One such numerical method is the Newton-Raphson algorithm.
+
+Our goal is to find the maximum likelihood estimate $\hat{\boldsymbol{\beta}}$.
+
+At $\hat{\boldsymbol{\beta}}$, the first derivative of the log-likelihood
+function will be equal to 0.
+
+Let's illustrate this by supposing
+
+$$
+\log \mathcal{L(\beta)} = - (\beta - 10) ^2 - 10
+$$
+
+Define the function `logL`.
+
+```{code-cell} ipython3
+@jax.jit
+def logL(β):
+    return -(β - 10) ** 2 - 10
+```
+
+
+To find the value of $\frac{d \log \mathcal{L(\boldsymbol{\beta})}}{d \boldsymbol{\beta}}$, we can use [jax.grad](https://jax.readthedocs.io/en/latest/_autosummary/jax.grad.html) which auto-differentiates the given function.
+
+We further use [jax.vmap](https://jax.readthedocs.io/en/latest/_autosummary/jax.vmap.html) which vectorizes the given function i.e. the function acting upon scalar inputs can now be used with vector inputs.
+
+```{code-cell} ipython3
+dlogL = jax.vmap(jax.grad(logL))
+```
+
+```{code-cell} ipython3
+β = jnp.linspace(1, 20)
+
+fig, (ax1, ax2) = plt.subplots(2, sharex=True, figsize=(12, 8))
+
+ax1.plot(β, logL(β), lw=2)
+ax2.plot(β, dlogL(β), lw=2)
+
+ax1.set_ylabel(r'$log \mathcal{L(\beta)}$',
+               rotation=0,
+               labelpad=35,
+               fontsize=15)
+ax2.set_ylabel(r'$\frac{dlog \mathcal{L(\beta)}}{d \beta}$ ',
+               rotation=0,
+               labelpad=35,
+               fontsize=19)
+
+ax2.set_xlabel(r'$\beta$', fontsize=15)
+ax1.grid(), ax2.grid()
+plt.axhline(c='black')
+plt.show()
+```
+
+
+The plot shows that the maximum likelihood value (the top plot) occurs
+when $\frac{d \log \mathcal{L(\boldsymbol{\beta})}}{d \boldsymbol{\beta}} = 0$ (the bottom
+plot).
+
+Therefore, the likelihood is maximized when $\beta = 10$.
+
+We can also ensure that this value is a *maximum* (as opposed to a
+minimum) by checking that the second derivative (slope of the bottom
+plot) is negative.
+
+The Newton-Raphson algorithm finds a point where the first derivative is
+0.
+
+To use the algorithm, we take an initial guess at the maximum value,
+$\beta_0$ (the OLS parameter estimates might be a reasonable
+guess), then
+
+
+Please refer to [this section](https://python.quantecon.org/mle.html#mle-with-numerical-methods) for the detailed algorithm.
+
+Let's have a go at implementing the Newton-Raphson algorithm.
+
+First, we'll create a `PoissonRegressionModel`.
+
+```{code-cell} ipython3
+PoissonRegressionModel = namedtuple('PoissonRegressionModel', ['X', 'y'])
+
+def create_poisson_model(X, y):
+    n, k = X.shape
+    # Reshape y as a n_by_1 column vector
+    y = y.reshape(n, 1)
+    X, y = jax.device_put((X, y))
+    return PoissonRegressionModel(X=X, y=y)
+```
+
+
+The following function computes the `factorial`.
+
+```{code-cell} ipython3
+@jax.jit
+def _factorial(n):
+    return jax.lax.exp(jax.lax.lgamma(n + 1.0)).astype(int)
+
```

Comment:
could you please provide a reference for this so the reader can understand what is happening?

Reply:
Added a small explanation above this cell. Please see if that's okay.